File: Populating Next Right Pointers in Each Node II.cpp
#include<iostream>


using namespace std;

struct TreeLinkNode {
    int val;
    TreeLinkNode *left, *right, *next;
    TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
};

TreeLinkNode* findnode(TreeLinkNode *root)
{
    if(root==NULL) return NULL;
    while(root->left==NULL&&root->right==NULL)
        {
            root=root->next;
            if(root==NULL) return NULL;
        }
    if(root->left!=NULL) return root->left;
    return root->right;
}

void connectTree(TreeLinkNode *root,TreeLinkNode *lchild,TreeLinkNode *rchild)
    {
        if(lchild!=NULL&&rchild!=NULL)
            {
                lchild->next=rchild;
                rchild->next=findnode(root->next);
                connectTree(rchild, rchild->left, rchild->right);
                connectTree(lchild, lchild->left, lchild->right);
            }
        else if(lchild!=NULL&&rchild==NULL)
            {
                lchild->next=findnode(root->next);
                connectTree(lchild, lchild->left, lchild->right);
            }
        else if(rchild!=NULL&&lchild==NULL)
            {
                rchild->next=findnode(root->next);
                connectTree(rchild, rchild->left, rchild->right);
            }
        return;
    }


void connect(TreeLinkNode *root)
    {
        if(root==NULL) return;
        connectTree(root, root->left, root->right);
    }

File: Python/Populating Next Right Pointers in Each Node II.py
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
        self.next = None

class Solution:
    # @param root, a tree node
    # @return nothing
    def findnextnode(self, root):
        if root == None:
            return None
        while root.left == None and root.right == None:
            root = root.next
            if root == None:
                return root
        if root.left != None:
            return root.left
        else:
            return root.right

    def connectTree(self, root, lchild, rchild):
        if lchild != None and rchild != None:
            lchild.next = rchild
            rchild.next = self.findnextnode(root.next)
            self.connectTree(rchild, rchild.left, rchild.right)
            self.connectTree(lchild, lchild.left, lchild.right)
        elif lchild != None and rchild == None:
            lchild.next = self.findnextnode(root.next)
            self.connectTree(lchild, lchild.left, lchild.right)
        elif rchild != None and lchild == None:
            rchild.next = self.findnextnode(root.next)
            self.connectTree(rchild, rchild.left, rchild.right)
        return

    def connect(self, root):
        if root == None:
            return
        self.connectTree(root, root.left, root.right)